### ISSUE TITLE
[Bug]: Searches Failing - fail to get shard leaders from QueryCoord: leader not available

### ISSUE THREAD

**Issue by IsaacWhittakerTR** [milvus-io/milvus#27936]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.3.1
- Deployment mode(standalone or cluster): Cluster (EKS)
- MQ type(rocksmq, pulsar or kafka): Kafka   
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus v2.3.1
- OS(Ubuntu or CentOS): Amazon Linux (EKS)
- CPU/Memory: 
- GPU: 
- Others: We do not have prometheus enabled yet
```

### Current Behavior

Sending any query to the cluster fails with the following error:

```failed to search: attempt #0: fail to get shard leaders from QueryCoord: leader not available: lastHB=2023-10-25 23:40:09.903580637 +0000 UTC: node=8: node offline: channel=by-dev-rootcoord-dml_1_444835380072571742v0: channel not available: attempt #1: fail to get shard leaders from QueryCoord: leader not available: lastHB=2023-10-25 23:40:09.903580637 +0000 UTC: node=8: node offline: channel=by-dev-rootcoord-dml_1_444835380072571742v0: channel not available: attempt #2: fail to get shard leaders from QueryCoord: leader not available: lastHB=2023-10-25 23:40:09.903580637 +0000 UTC: node=8: node offline: channel=by-dev-rootcoord-dml_1_444835380072571742v0: channel not available: attempt #3: fail to get shard leaders from QueryCoord: leader not available: lastHB=2023-10-25 23:40:09.903580637 +0000 UTC: node=8: node offline: channel=by-dev-rootcoord-dml_1_444835380072571742v0: channel not available: attempt #4: fail to get shard leaders from QueryCoord: leader not available: lastHB=2023-10-25 23:40:09.903580637 +0000 UTC: node=8: node offline: channel=by-dev-rootcoord-dml_1_444835380072571742v0: channel not available: attempt #5: fail to get shard leaders from QueryCoord: leader not available: lastHB=2023-10-25 23:40:09.903580637 +0000 UTC: node=8: node offline: channel=by-dev-rootcoord-dml_1_444835380072571742v0: channel not available: attempt #6: fail to get shard leaders from QueryCoord: leader not available: lastHB=2023-10-25 23:40:09.903580637 +0000 UTC: node=8: node offline: channel=by-dev-rootcoord-dml_1_444835380072571742v0: channel not available: context done during sleep after run#6: context deadline exceeded```

This started occurring suddenly with an existing Collection which had previously been working correctly. I also tried creating a new Collection and still get the same errors with the new collection.

All k8s pods are healthy and did not restart.

### Expected Behavior

Queries should return a response without error.

### Steps To Reproduce

```markdown
Repro steps:

1. Create a new collection using `pymilvus`
2. Create new index (type=`IVF_SQ8`, metric_type=`L2`, nlist=`16384`, uses partition key with num_partitions=`4000`)
3. Upload ~3000 entities to the collection
4. Load the collection into memory
5. Query the collection -> This is where the error occurs
```

### Milvus Log

See this log file for the recent ~10 minutes of logs after I created a new Collection and reproduced the query errors.

[bug-searches-failing-milvus-log.tar.gz](https://github.com/milvus-io/milvus/files/13171783/bug-searches-failing-milvus-log.tar.gz)

### Anything else?

After doing some troubleshooting on the original bug, I noticed some additional error logs on the index nodes related to Minio (we are using s3). But I do not know if those warning/error messages are related to the query-time bug, or an unrelated issue.. any advice on which error logs to focus on or what troubleshooting steps I can take would be appreciated.

Example warnings:
```[2023/10/25 23:46:38.750 +00:00] [WARN] [indexcgowrapper/helper.go:78] ["failed to create index, C Runtime Exception: [UnexpectedError] Error:GetObjectSize[errcode:400, exception:, errmessage:No response body.]\n"]
[2023/10/25 23:46:38.750 +00:00] [ERROR] [indexnode/task.go:347] ["failed to build index"] [error="[UnexpectedError] Error:GetObjectSize[errcode:400, exception:, errmessage:No response body.]"] [errorVerbose="[UnexpectedError] Error:GetObjectSize[errcode:400, exception:, errmessage:No response body.]\n(1) attached stack trace\n  -- stack trace:\n  | github.com/milvus-io/milvus/internal/util/indexcgowrapper.HandleCStatus\n  | \t/go/src/github.com/milvus-io/milvus/internal/util/indexcgowrapper/helper.go:79\n  | github.com/milvus-io/milvus/internal/util/indexcgowrapper.CreateIndex\n  | \t/go/src/github.com/milvus-io/milvus/internal/util/indexcgowrapper/index.go:100\n  | github.com/milvus-io/milvus/internal/indexnode.(*indexBuildTask).BuildIndex\n  | \t/go/src/github.com/milvus-io/milvus/internal/indexnode/task.go:340\n  | github.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask.func1\n  | \t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:209\n  | github.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask\n  | \t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:222\n  | github.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).indexBuildLoop.func1\n  | \t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:255\n  | runtime.goexit\n  | \t/usr/local/go/src/runtime/asm_amd64.s:1598\nWraps: (2) [UnexpectedError] Error:GetObjectSize[errcode:400, exception:, errmessage:No response body.]\nError types: (1) *withstack.withStack (2) *errutil.leafError"] [stack="github.com/milvus-io/milvus/internal/indexnode.(*indexBuildTask).BuildIndex\n\t/go/src/github.com/milvus-io/milvus/internal/indexnode/task.go:347\ngithub.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask.func1\n\t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:209\ngithub.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask\n\t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:222\ngithub.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).indexBuildLoop.func1\n\t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:255"]
```

**Comment by yhmo** [comment]
[2023/10/25 17:52:30.464 +00:00] [WARN] [delegator/delegator_data.go:308] ["failed to load growing segment"] [traceID=b36293bfd89bdb52d831b1cf62118a70] [collectionID=444835380068840090] [channel=by-dev-rootcoord-dml_0_444835380068840090v0] [replicaID=444835382227894275] [error="limit=160: request limit exceeded"]

Possible reason: segment has many data files, parallel reading threads hit the rate limit of aws s3.

A workaround, change a configuration to decrease the parallel reading threads.
In the milvus.yaml change the `common.threadCoreCoefficient.highPriority` value from 10 to a small value(2 or 3).
The number of parallel reading threads is calculated by `common.threadCoreCoefficient.highPriority * querynode_cpu_cores`
![Screenshot from 2023-10-26 17-57-07](https://github.com/milvus-io/milvus/assets/2282099/179123bd-f8f0-44bd-aaae-e795faab2e07)

**Comment by IsaacWhittakerTR** [comment]
Thanks for the prompt response, I will definitely try updating the `common.threadCoreCoefficient.highPriority` configuration.

Your observation also got me looking more into the data on s3 and I noticed that the `index_files/` object (folder) on s3 is completely missing for this cluster. Also, the memory usage for the kafka nodes is quite high.. which leads me to believe that the cluster has not been able to write any index files to s3 since it was created.

We manually confirmed that the credentials we are using have the correct permissions to PutObject and GetObject for this bucket and ListBucket for all buckets (and the delta_log, insert_log, and stats_log all exist and seem to be working fine).

This is our first time deploying a brand new cluster on v2.3.1, but we have another cluster that was upgraded and is working fine- Do you know if there were any recent updates (between Milvus v2.2.7 and v2.3.1) which would change the way that Kafka works, or that would change the way the index segments are written to storage when using s3?

**Comment by IsaacWhittakerTR** [comment]
For example would there have been any change as part of the PR associated with [this issue](https://github.com/milvus-io/milvus/issues/27509) that could be causing our cluster to fail to write the index files to s3?

**Comment by IsaacWhittakerTR** [comment]
I found this issue which had similar error messages reported: [[Bug]: S3 Milvus - can insert entities but failing to create index (Error:GetObjectSize: No response body.) #25939](https://github.com/milvus-io/milvus/issues/25939). 

Based on one of the suggestions there, I tried adding the configuration `externalS3.region` to our yaml file. However to do so I had to use a more recent version of the helm chart, v4.1.7 which I believe corresponds with the new Milvus v2.3.1 release (I had to upgrade because that field was not added until [this PR](https://github.com/zilliztech/milvus-helm/pull/13) created 3 weeks ago).

This seemed to have helped, because suddenly the cluster was able to write files to the `index_files/` object on s3 again. However I started getting new errors like this:

`[2023/10/29 02:59:14.351 +00:00] [INFO] [indexnode/task.go:184] ["Successfully prepare indexBuildTask"] [buildID=444835380072508708] [Collection=0] [SegmentID=0]
[2023/10/29 02:59:14.375 +00:00] [INFO] [indexnode/task.go:325] ["index params are ready"] [buildID=444835380072508708] ["index params"="{\"dim\":\"768\",\"index_type\":\"IVF_SQ8\",\"max_length\":\"100\",\"metric_type\":\"L2\",\"nlist\":\"16384\"}"]
E20231029 02:59:14.446923  52 config.cc:65] [KNOWHERE][FormatAndCheck][milvus] invalid json key max_length
I20231029 02:59:14.375972  52 MinioChunkManager.cpp:114] [SEGCORE][InitSDKAPI][milvus] init aws with log level:warn
I20231029 02:59:14.390561  52 MinioChunkManager.cpp:85] [SEGCORE][ProcessFormattedStatement][milvus] [AWS LOG] [WARN] 2023-10-29 02:59:14.390 ClientConfiguration [140089025881856] Retry Strategy will use the default max attempts.
I20231029 02:59:14.390647  52 MinioChunkManager.cpp:321] [SEGCORE][MinioChunkManager][milvus] init MinioChunkManager with parameter[endpoint: '[s3.us-west-2.amazonaws.com:80](http://s3.us-west-2.amazonaws.com/)', default_bucket_name:'spaitr-milvus-storage-stg', root_path:'', use_secure:'false']
I20231029 02:59:14.390676  52 factory.cc:20] [KNOWHERE][Create][milvus] create knowhere index IVF_SQ8 with version 1
 => failed to build index, invalid param in json
[2023/10/29 02:59:14.447 +00:00] [WARN] [indexcgowrapper/helper.go:71] ["failed to create index, C Runtime Exception: => failed to build index, invalid param in json\n"]
[2023/10/29 02:59:14.448 +00:00] [ERROR] [indexnode/task.go:356] ["failed to build index"] [error="failed to create index, C Runtime Exception: => failed to build index, invalid param in json\n: internal code=2001: segcore error"] [errorVerbose="failed to create index, C Runtime Exception: => failed to build index, invalid param in json: internal code=2001: segcore error\n(1) attached stack trace\n -- stack trace:\n | [github.com/milvus-io/milvus/pkg/util/merr.WrapErrSegcore](http://github.com/milvus-io/milvus/pkg/util/merr.WrapErrSegcore)\n | \t/go/src/github.com/milvus-io/milvus/pkg/util/merr/utils.go:737\n | [...repeated from below...]\nWraps: (2) failed to create index, C Runtime Exception: => failed to build index, invalid param in json\nWraps: (3) attached stack trace\n -- stack trace:\n | [github.com/milvus-io/milvus/pkg/util/merr.WrapErrSegcore](http://github.com/milvus-io/milvus/pkg/util/merr.WrapErrSegcore)\n | \t/go/src/github.com/milvus-io/milvus/pkg/util/merr/utils.go:735\n | [github.com/milvus-io/milvus/internal/util/indexcgowrapper.HandleCStatus](http://github.com/milvus-io/milvus/internal/util/indexcgowrapper.HandleCStatus)\n | \t/go/src/github.com/milvus-io/milvus/internal/util/indexcgowrapper/helper.go:72\n | [github.com/milvus-io/milvus/internal/util/indexcgowrapper.CreateIndex](http://github.com/milvus-io/milvus/internal/util/indexcgowrapper.CreateIndex)\n | \t/go/src/github.com/milvus-io/milvus/internal/util/indexcgowrapper/index.go:98\n | [github.com/milvus-io/milvus/internal/indexnode.(*indexBuildTask).BuildIndex](http://github.com/milvus-io/milvus/internal/indexnode.(*indexBuildTask).BuildIndex)\n | \t/go/src/github.com/milvus-io/milvus/internal/indexnode/task.go:349\n | [github.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask.func1](http://github.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask.func1)\n | \t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:209\n | [github.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask](http://github.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask)\n | \t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:222\n | [github.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).indexBuildLoop.func1](http://github.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).indexBuildLoop.func1)\n | \t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:255\n | runtime.goexit\n | \t/usr/local/go/src/runtime/asm_amd64.s:1598\nWraps: (4) internal code=2001\nWraps: (5) segcore error\nError types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) merr.milvusError"] [stack="[github.com/milvus-io/milvus/internal/indexnode.(*indexBuildTask).BuildIndex](http://github.com/milvus-io/milvus/internal/indexnode.(*indexBuildTask).BuildIndex)\n\t/go/src/github.com/milvus-io/milvus/internal/indexnode/task.go:356\[ngithub.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask.func1](http://ngithub.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask.func1)\n\t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:209\[ngithub.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask](http://ngithub.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).processTask)\n\t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:222\[ngithub.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).indexBuildLoop.func1](http://ngithub.com/milvus-io/milvus/internal/indexnode.(*TaskScheduler).indexBuildLoop.func1)\n\t/go/src/github.com/milvus-io/milvus/internal/indexnode/task_scheduler.go:255"]
[2023/10/29 02:59:16.306 +00:00] [INFO] [indexnode/indexnode_service.go:173] ["drop index build jobs"] [traceID=81be980d2657ae9947ca4d4fb71f60db] [clusterID=by-dev] [indexBuildIDs="[444835380072508708]"]`

Any ideas what could be causing this issue?
When the error message says `failed to build index, invalid param in json` what JSON is that referring to?
I also see an error log referencing `max_length`, could that be referring to the max_length I set for some of the fields in our schema?

**Comment by IsaacWhittakerTR** [comment]
I found this issue which had similar error messages reported: [[Bug]: S3 Milvus - can insert entities but failing to create index (Error:GetObjectSize: No response body.) #25939](https://github.com/milvus-io/milvus/issues/25939).

Based on one of the suggestions there, I tried adding the configuration `externalS3.region` to our yaml file. However to do so I had to use a more recent version of the helm chart, `v4.1.7` which I believe corresponds with the new Milvus `v2.3.1` release (I had to upgrade because that field was not added until [this PR](https://github.com/zilliztech/milvus-helm/pull/13) created 3 weeks ago).

This seems to have possibly fixed the issue, because suddenly the cluster was able to write files to the `index_files/` object on s3 again. However I still see the kafka nodes are using quite a bit of memory, and it is hard to tell if all the data has been fully indexed.

Is there any way to confirm that all the index segments have been created for a certain collection?
